sound/soc/intel/boards/sof_sdw.c

```diff
@@ -515,6 +515,47 @@ int sdw_trigger(struct snd_pcm_substream *substream, int cmd)
 	return ret;
 }
 
+int sdw_hw_params(struct snd_pcm_substream *substream,
+		  struct snd_pcm_hw_params *params)
+{
+	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
+	struct snd_soc_dai *cpu_dai;
+	struct snd_soc_dai *codec_dai;
+	unsigned int ch_mask;
+	int num_codecs;
+	int step;
+	int i;
+	int j;
+
+	if (!rtd->dai_link->codec_ch_maps)
+		return 0;
+
+	/* Identical data will be sent to all codecs in playback */
+	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK) {
+		step = 0;
+		ch_mask = GENMASK(params_channels(params) - 1, 0);
+	} else {
+		num_codecs = rtd->dai_link->num_codecs;
+		ch_mask = GENMASK(params_channels(params) / num_codecs - 1, 0);
```

Comment:
The core is following what we are telling to it via the `ch_mask`, so we need to put all the logic and safeguarding here, right?

What about a case when you have let's say 4 codecs (each stereo) and a user opens a stereo capture?
`2 / 4 - 1 = -1` ?
What about  a case when you have different codecs, two codec one mono and one 4 channel?
I guess this is an exotic setup, but at least we should fail it atm.

The other case is 2x stereo codecs and a 3 channel capture, what to do?

We should ensure that the codecs have identical number of channels supported, the channel count step should be constrained to codec_channels step.

Reply:
Hmm, I added sanity check, but I will let component driver check if the channel number is supported. Currently, we split channel evenly, and will return error if we can't.